Change Builder type parameters to support trait objects on builder methods

Howdy. 

I feel like it should be possible to change the function signature on the field methods to use trait objects. These are monomorphic at compile time and wouldn't have any runtime penalty compared to writing the code by hand.

Examples:

[Into](https://doc.rust-lang.org/std/convert/trait.Into.html) for owned valued:
```Rust
pub fn bar(self, baz: String) -> ReturnType
```

```Rust
pub fn bar(self, baz: impl Into<String>) -> ReturnType
```
[AsRef](https://doc.rust-lang.org/std/convert/trait.AsRef.html) for References

```Rust
pub fn bar(self, baz: &str) -> ReturnType
```
```Rust
pub fn bar(self, baz: impl AsRef<str>) -> ReturnType
```
[AsMut](https://doc.rust-lang.org/std/convert/trait.AsMut.html) for mutable references

```Rust
pub fn bar(self, baz: &mut str) -> ReturnType
```
```Rust
pub fn bar(self, baz: impl AsMut<str>) -> ReturnType
```

This would allow any type that could be represented as the type the builder needs to be used as a satisfactory parameter.

Example:

Explicit Type Parameter
```Rust
let foo = Foo::builder()
    .bar("Hello".to_string())
    .baz("world!".to_string())
    .qux("Hello".to_string())
    .quz("world!".to_string())
    .build();
```
With `impl Trait` Type Parameter
```Rust
let foo = Foo::builder()
    .bar("Hello")
    .baz("world!")
    .qux("Hello")
    .quz("world!")
    .build();
```

This is often more flexible and less verbose as a public API since you don't *have* to provide a string, just something that could be represented as a string.

@MinisculeGirraffe I'm having some trouble implementing `AsRef`, I hoped you could help me out. `Into` was very easy, and I haven't started with `AsMut` yet. 

The problem lies within the setter. I'm basically having issues with the lifetime.

For now, if I have this struct
```rust
#[derive(Builder)]
pub struct Foo<'a> {
   #[builder(setter = asref)]
    bar: &'a str,
}
```

It will desugar into:
```rust
pub struct Foo<'a> {
    bar: &'a str,
}

pub trait Builder<'a> {
    type BuilderImpl;
    fn builder() -> Self::BuilderImpl;
}

impl<'a> Builder<'a> for Foo<'a> {
    type BuilderImpl = FooBuilder<'a, false>;
    #[doc = "Creates an instance of [`FooBuilder`]"]
    fn builder() -> Self::BuilderImpl {
        Self::BuilderImpl::new()
    }
}
#[doc = "Builder for [`Foo`] derived using the `const_typed_builder` crate"]
pub struct FooBuilder<'a, const __BUILDER_CONST_0: bool> {
    __foo_data: FooData<'a>,
}
impl<'a> FooBuilder<'a, false> {
    #[doc = "Creates a new [`FooBuilder`] without any fields initialized"]
    pub fn new() -> FooBuilder<'a, false> {
        Self::default()
    }
}
impl<'a> Default for FooBuilder<'a, false> {
    #[doc = "Creates a new [`FooBuilder`] without any fields initialized"]
    fn default() -> Self {
        FooBuilder {
            __foo_data: FooData::default(),
        }
    }
}
impl<'a> FooBuilder<'a, false> {
    #[doc = "\nSetter for the [`Foo::bar`] field.\n\n# Arguments\n\n- `bar`: field to be set\n\n# Returns\n\n`FooBuilder` with `bar` initialized"]
    pub fn bar(self, bar: impl AsRef<str> + 'a) -> FooBuilder<'a, true> {
        let mut __foo_data = self.__foo_data;
        __foo_data.bar = Some(bar.as_ref());
        FooBuilder { __foo_data } // this line does not work
    }
}
impl<'a> FooBuilder<'a, true> {
    #[doc = "Build an instance of [`Foo`], consuming the [`FooBuilder`]"]
    pub fn build(self) -> Foo<'a> {
        self.__foo_data.into()
    }
}
#[doc(hidden)]
pub struct FooData<'a> {
    pub bar: Option<&'a str>,
}
impl<'a> From<FooData<'a>> for Foo<'a> {
    #[doc(hidden)]
    fn from(data: FooData<'a>) -> Foo<'a> {
        Foo {
            bar: data.bar.unwrap(),
        }
    }
}
impl<'a> Default for FooData<'a> {
    #[doc(hidden)]
    fn default() -> Self {
        FooData { bar: None }
    }
}
```
But that gives me the following error:
```
error[E0515]: cannot return value referencing function parameter `bar`
  --> src/main.rs:40:9
   |
39 |         __foo_data.bar = Some(bar.as_ref());
   |                               --- `bar` is borrowed here
40 |         FooBuilder { __foo_data } // this line does not work
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^ returns a value referencing data owned by the current function
```

I also thought about using `std::mem::transmute`, even though I'm hesitant to use `unsafe`
```rust
impl<'a> FooBuilder<'a, false> {
    pub fn bar(mut self, bar: impl AsRef<str> + 'a) -> FooBuilder<'a, true> {
        self.__foo_data.bar = Some(bar.as_ref());
        unsafe { std::mem::transmute(self) }
    }
}
```

```
error[E0597]: `bar` does not live long enough
  --> src/main.rs:38:36
   |
35 | impl<'a> FooBuilder<'a, false> {
   |      -- lifetime `'a` defined here
36 |     #[doc = "\nSetter for the [`Foo::bar`] field.\n\n# Arguments\n\n- `bar...
37 |     pub fn bar(mut self, bar: impl AsRef<str> + 'a) -> FooBuilder<'a, true> {
   |                          --- binding `bar` declared here
38 |         self.__foo_data.bar = Some(bar.as_ref());
   |         ---------------------------^^^----------
   |         |                          |
   |         |                          borrowed value does not live long enough
   |         assignment requires that `bar` is borrowed for `'a`
39 |         unsafe { std::mem::transmute(self) }
40 |     }
   |     - `bar` dropped here while still borrowed
```

--- a/const_typed_builder_derive/src/generator/builder_generator.rs
+++ b/const_typed_builder_derive/src/generator/builder_generator.rs
@@ -1,6 +1,6 @@
 use super::util;
 use crate::info::{
-    Container, Field, FieldKind, GroupType, SolverKind, TrackedField, TrackedFieldKind, SetterKind,
+    Container, Field, FieldKind, GroupType, SetterKind, SolverKind, TrackedField, TrackedFieldKind,
 };
 use itertools::{Itertools, Powerset};
 use proc_macro2::TokenStream;
@@ -425,12 +425,13 @@ Setter for the [`{}::{field_ident}`] field.
         }
 
         let input_type = match field.setter_kind() {
-            SetterKind::Standard => {
-                match field.kind() {
-                    FieldKind::Grouped | FieldKind::Mandatory if field.is_option_type() => field.inner_type().expect("Option types have an inner type").to_token_stream(),
-                    FieldKind::Skipped => unreachable!("Skipped fields have an early return"),
-                    _ => field.ty().to_token_stream()
-                }
+            SetterKind::Standard => match field.kind() {
+                FieldKind::Grouped | FieldKind::Mandatory if field.is_option_type() => field
+                    .inner_type()
+                    .expect("Option types have an inner type")
+                    .to_token_stream(),
+                FieldKind::Skipped => unreachable!("Skipped fields have an early return"),
+                _ => field.ty().to_token_stream(),
             },
             SetterKind::Propagate => {
                 let input = if let Some(inner_ty) = field.inner_type() {
@@ -439,12 +440,14 @@ Setter for the [`{}::{field_ident}`] field.
                     field.ty()
                 };
                 let output = match field.kind() {
-                    FieldKind::Grouped | FieldKind::Mandatory if field.is_option_type() => field.inner_type().expect("Option types have an inner type"),
+                    FieldKind::Grouped | FieldKind::Mandatory if field.is_option_type() => {
+                        field.inner_type().expect("Option types have an inner type")
+                    }
                     FieldKind::Skipped => unreachable!("Skipped fields have an early return"),
-                    _ => field.ty()
+                    _ => field.ty(),
                 };
                 quote!(fn(<#input as Builder>:: BuilderImpl) -> #output)
-            },
+            }
             SetterKind::Into => {
                 let ty = if let Some(inner_ty) = field.inner_type() {
                     inner_ty
@@ -456,34 +459,52 @@ Setter for the [`{}::{field_ident}`] field.
                 } else {
                     quote!(impl Into<#ty>)
                 }
-            },
+            }
             SetterKind::AsMut => {
                 let ty = if let Some(inner_ty) = field.inner_type() {
                     inner_ty
                 } else {
-                    field.ty() 
+                    field.ty()
                 };
-                if let syn::Type::Reference(syn::TypeReference { lifetime, mutability, elem, ..}) = ty {
+                if let syn::Type::Reference(syn::TypeReference {
+                    lifetime,
+                    mutability,
+                    elem,
+                    ..
+                }) = ty
+                {
                     if mutability.is_none() {
-                        emit_error!(ty, "You need a mutable reference to use this type of setter");
+                        emit_error!(
+                            ty,
+                            "You need a mutable reference to use this type of setter"
+                        );
                     }
                     if field.kind() == FieldKind::Optional {
                         quote!(Option<&#lifetime mut impl AsMut<#elem>>)
-                    } else { 
+                    } else {
                         quote!(&#lifetime mut impl AsMut<#elem>)
                     }
                 } else {
-                    emit_error!(ty, "You need a mutable reference to use this type of setter");
+                    emit_error!(
+                        ty,
+                        "You need a mutable reference to use this type of setter"
+                    );
                     return None;
                 }
-            },
+            }
             SetterKind::AsRef => {
                 let ty = if let Some(inner_ty) = field.inner_type() {
                     inner_ty
                 } else {
-                    field.ty() 
+                    field.ty()
                 };
-                if let syn::Type::Reference(syn::TypeReference { lifetime, mutability: _, elem, .. }) = ty {
+                if let syn::Type::Reference(syn::TypeReference {
+                    lifetime,
+                    mutability: _,
+                    elem,
+                    ..
+                }) = ty
+                {
                     if field.kind() == FieldKind::Optional {
                         quote!(Option<&#lifetime impl AsRef<#elem>>)
                     } else {
@@ -493,7 +514,7 @@ Setter for the [`{}::{field_ident}`] field.
                     emit_error!(ty, "You need a reference to use this type of setter");
                     return None;
                 }
-            },
+            }
         };
         Some(input_type)
     }
@@ -512,27 +533,35 @@ Setter for the [`{}::{field_ident}`] field.
                 } else {
                     quote!(Some(#field_ident))
                 }
-            },
-            SetterKind::Propagate => if let Some(inner_ty) = field.inner_type() {
-                quote!(#field_ident(<#inner_ty as Builder>::builder()))
-            } else {
-                let ty = field.ty();
-                quote!(Some(#field_ident(<#ty as Builder>::builder())))
             }
-            SetterKind::Into => if field.kind() == FieldKind::Optional {
-                quote!(#field_ident.map(Into::into))
-            } else {
-                quote!(Some(#field_ident.into()))
-            },
-            SetterKind::AsMut => if field.kind() == FieldKind::Optional {
-                quote!(#field_ident.map(AsMut::as_mut))
-            } else {
-                quote!(Some(#field_ident.as_mut()))
-            },
-            SetterKind::AsRef => if field.kind() == FieldKind::Optional {
-                quote!(#field_ident.map(AsRef::as_ref))
-            } else {
-                quote!(Some(#field_ident.as_ref()))
+            SetterKind::Propagate => {
+                if let Some(inner_ty) = field.inner_type() {
+                    quote!(#field_ident(<#inner_ty as Builder>::builder()))
+                } else {
+                    let ty = field.ty();
+                    quote!(Some(#field_ident(<#ty as Builder>::builder())))
+                }
+            }
+            SetterKind::Into => {
+                if field.kind() == FieldKind::Optional {
+                    quote!(#field_ident.map(Into::into))
+                } else {
+                    quote!(Some(#field_ident.into()))
+                }
+            }
+            SetterKind::AsMut => {
+                if field.kind() == FieldKind::Optional {
+                    quote!(#field_ident.map(AsMut::as_mut))
+                } else {
+                    quote!(Some(#field_ident.as_mut()))
+                }
+            }
+            SetterKind::AsRef => {
+                if field.kind() == FieldKind::Optional {
+                    quote!(#field_ident.map(AsRef::as_ref))
+                } else {
+                    quote!(Some(#field_ident.as_ref()))
+                }
             }
         };
 

--- a/const_typed_builder_derive/src/generator/builder_generator.rs
+++ b/const_typed_builder_derive/src/generator/builder_generator.rs
@@ -179,7 +179,7 @@ impl<'info> BuilderGenerator<'info> {
             .info
             .field_collection()
             .iter()
-            .filter(|field| field.kind() != &FieldKind::Skipped)
+            .filter(|field| field.kind() != FieldKind::Skipped)
             .map(|field| {
                 let const_idents_impl = self.const_generic_idents_set_impl(field);
                 let const_idents_type_input = self.const_generic_idents_set_type(field, false);
@@ -418,28 +418,35 @@ Setter for the [`{}::{field_ident}`] field.
     }
 
     fn impl_set_input_type(&self, field: &'info Field) -> Option<TokenStream> {
-        if field.kind() == &FieldKind::Skipped {
-            return None;
+        match field.setter_kind() {
+            crate::info::SetterKind::Standard => {
+                if field.kind() == FieldKind::Skipped {
+                    return None;
+                }
+                let field_ident = field.ident();
+                let field_ty = field.setter_input_type();
+                let bottom_ty = if field.is_option_type() {
+                    field.inner_type().unwrap()
+                } else {
+                    field_ty.unwrap()
+                };
+        
+                let field_ty = if field.propagate() {
+                    quote!(fn(<#bottom_ty as Builder>:: BuilderImpl) -> #field_ty)
+                } else {
+                    quote!(#field_ty)
+                };
+        
+                Some(quote!(#field_ident: #field_ty))
+            },
+            crate::info::SetterKind::Into => todo!(),
+            crate::info::SetterKind::AsMut => todo!(),
+            crate::info::SetterKind::AsRef => todo!(),
         }
-        let field_ident = field.ident();
-        let field_ty = field.setter_input_type();
-        let bottom_ty = if field.is_option_type() {
-            field.inner_type().unwrap()
-        } else {
-            field_ty.unwrap()
-        };
-
-        let field_ty = if field.propagate() {
-            quote!(fn(<#bottom_ty as Builder>:: BuilderImpl) -> #field_ty)
-        } else {
-            quote!(#field_ty)
-        };
-
-        Some(quote!(#field_ident: #field_ty))
     }
 
     fn impl_set_input_value(&self, field: &'info Field) -> Option<TokenStream> {
-        if field.kind() == &FieldKind::Skipped {
+        if field.kind() == FieldKind::Skipped {
             return None;
         }
 
@@ -457,7 +464,7 @@ Setter for the [`{}::{field_ident}`] field.
             quote!(#field_ident)
         };
 
-        if field.kind() == &FieldKind::Optional {
+        if field.kind() == FieldKind::Optional {
             Some(quote!(#field_value))
         } else {
             Some(quote!(Some(#field_value)))

--- a/const_typed_builder_derive/src/generator/builder_generator.rs
+++ b/const_typed_builder_derive/src/generator/builder_generator.rs
@@ -1,10 +1,11 @@
 use super::util;
 use crate::info::{
-    Container, Field, FieldKind, GroupType, SolverKind, TrackedField, TrackedFieldKind,
+    Container, Field, FieldKind, GroupType, SolverKind, TrackedField, TrackedFieldKind, SetterKind,
 };
 use itertools::{Itertools, Powerset};
 use proc_macro2::TokenStream;
-use quote::quote;
+use proc_macro_error::emit_error;
+use quote::{quote, ToTokens};
 use std::{collections::BTreeSet, ops::Deref};
 use syn::{parse_quote, GenericParam};
 
@@ -187,8 +188,8 @@ impl<'info> BuilderGenerator<'info> {
                 let where_clause = &self.info.generics().where_clause;
 
                 let field_ident = field.ident();
-                let input_type = self.impl_set_input_type(field);
-                let input_value = self.impl_set_input_value(field);
+                let input_type = self.field_input_type(field);
+                let input_value = self.field_input_assign(field);
 
                 let documentation = format!(r#"
 Setter for the [`{}::{field_ident}`] field.
@@ -204,7 +205,7 @@ Setter for the [`{}::{field_ident}`] field.
                 quote!(
                     impl #const_idents_impl #builder_ident #const_idents_type_input #where_clause {
                         #[doc = #documentation]
-                        pub fn #field_ident (self, #input_type) -> #builder_ident #const_idents_type_output {
+                        pub fn #field_ident (self, #field_ident: #input_type) -> #builder_ident #const_idents_type_output {
                             let mut #data_field = self.#data_field;
                             #data_field.#field_ident = #input_value;
                             #builder_ident {
@@ -417,58 +418,125 @@ Setter for the [`{}::{field_ident}`] field.
         )
     }
 
-    fn impl_set_input_type(&self, field: &'info Field) -> Option<TokenStream> {
-        match field.setter_kind() {
-            crate::info::SetterKind::Standard => {
-                if field.kind() == FieldKind::Skipped {
+    /// Retrieves the input type for the builder's setter method.
+    pub fn field_input_type(&self, field: &Field) -> Option<TokenStream> {
+        if field.kind() == FieldKind::Skipped {
+            return None;
+        }
+
+        let input_type = match field.setter_kind() {
+            SetterKind::Standard => {
+                match field.kind() {
+                    FieldKind::Grouped | FieldKind::Mandatory if field.is_option_type() => field.inner_type().expect("Option types have an inner type").to_token_stream(),
+                    FieldKind::Skipped => unreachable!("Skipped fields have an early return"),
+                    _ => field.ty().to_token_stream()
+                }
+            },
+            SetterKind::Propagate => {
+                let input = if let Some(inner_ty) = field.inner_type() {
+                    inner_ty
+                } else {
+                    field.ty()
+                };
+                let output = match field.kind() {
+                    FieldKind::Grouped | FieldKind::Mandatory if field.is_option_type() => field.inner_type().expect("Option types have an inner type"),
+                    FieldKind::Skipped => unreachable!("Skipped fields have an early return"),
+                    _ => field.ty()
+                };
+                quote!(fn(<#input as Builder>:: BuilderImpl) -> #output)
+            },
+            SetterKind::Into => {
+                let ty = if let Some(inner_ty) = field.inner_type() {
+                    inner_ty
+                } else {
+                    field.ty()
+                };
+                if field.kind() == FieldKind::Optional {
+                    quote!(Option<impl Into<#ty>>)
+                } else {
+                    quote!(impl Into<#ty>)
+                }
+            },
+            SetterKind::AsMut => {
+                let ty = if let Some(inner_ty) = field.inner_type() {
+                    inner_ty
+                } else {
+                    field.ty() 
+                };
+                if let syn::Type::Reference(syn::TypeReference { lifetime, mutability, elem, ..}) = ty {
+                    if mutability.is_none() {
+                        emit_error!(ty, "You need a mutable reference to use this type of setter");
+                    }
+                    if field.kind() == FieldKind::Optional {
+                        quote!(Option<&#lifetime mut impl AsMut<#elem>>)
+                    } else { 
+                        quote!(&#lifetime mut impl AsMut<#elem>)
+                    }
+                } else {
+                    emit_error!(ty, "You need a mutable reference to use this type of setter");
                     return None;
                 }
-                let field_ident = field.ident();
-                let field_ty = field.setter_input_type();
-                let bottom_ty = if field.is_option_type() {
-                    field.inner_type().unwrap()
-                } else {
-                    field_ty.unwrap()
-                };
-        
-                let field_ty = if field.propagate() {
-                    quote!(fn(<#bottom_ty as Builder>:: BuilderImpl) -> #field_ty)
-                } else {
-                    quote!(#field_ty)
-                };
-        
-                Some(quote!(#field_ident: #field_ty))
             },
-            crate::info::SetterKind::Into => todo!(),
-            crate::info::SetterKind::AsMut => todo!(),
-            crate::info::SetterKind::AsRef => todo!(),
-        }
+            SetterKind::AsRef => {
+                let ty = if let Some(inner_ty) = field.inner_type() {
+                    inner_ty
+                } else {
+                    field.ty() 
+                };
+                if let syn::Type::Reference(syn::TypeReference { lifetime, mutability: _, elem, .. }) = ty {
+                    if field.kind() == FieldKind::Optional {
+                        quote!(Option<&#lifetime impl AsRef<#elem>>)
+                    } else {
+                        quote!(&#lifetime impl AsRef<#elem>)
+                    }
+                } else {
+                    emit_error!(ty, "You need a reference to use this type of setter");
+                    return None;
+                }
+            },
+        };
+        Some(input_type)
     }
 
-    fn impl_set_input_value(&self, field: &'info Field) -> Option<TokenStream> {
+    pub fn field_input_assign(&self, field: &Field) -> Option<TokenStream> {
         if field.kind() == FieldKind::Skipped {
             return None;
         }
 
         let field_ident = field.ident();
-        let field_ty = field.setter_input_type();
-        let bottom_ty = if field.is_option_type() {
-            field.inner_type().unwrap()
-        } else {
-            field_ty.unwrap()
+
+        let field_value = match field.setter_kind() {
+            SetterKind::Standard => {
+                if field.kind() == FieldKind::Optional {
+                    quote!(#field_ident)
+                } else {
+                    quote!(Some(#field_ident))
+                }
+            },
+            SetterKind::Propagate => if let Some(inner_ty) = field.inner_type() {
+                quote!(#field_ident(<#inner_ty as Builder>::builder()))
+            } else {
+                let ty = field.ty();
+                quote!(Some(#field_ident(<#ty as Builder>::builder())))
+            }
+            SetterKind::Into => if field.kind() == FieldKind::Optional {
+                quote!(#field_ident.map(Into::into))
+            } else {
+                quote!(Some(#field_ident.into()))
+            },
+            SetterKind::AsMut => if field.kind() == FieldKind::Optional {
+                quote!(#field_ident.map(AsMut::as_mut))
+            } else {
+                quote!(Some(#field_ident.as_mut()))
+            },
+            SetterKind::AsRef => if field.kind() == FieldKind::Optional {
+                quote!(#field_ident.map(AsRef::as_ref))
+            } else {
+                quote!(Some(#field_ident.as_ref()))
+            }
         };
 
-        let field_value = if field.propagate() {
-            quote!(#field_ident(<#bottom_ty as Builder>::builder()))
-        } else {
-            quote!(#field_ident)
-        };
-
-        if field.kind() == FieldKind::Optional {
-            Some(quote!(#field_value))
-        } else {
-            Some(quote!(Some(#field_value)))
-        }
+        Some(quote!(#field_value))
     }
 
     fn valid_groupident_combinations(&self) -> impl Iterator<Item = Vec<usize>> + '_ {

--- a/const_typed_builder_derive/src/generator/builder_generator.rs
+++ b/const_typed_builder_derive/src/generator/builder_generator.rs
@@ -205,11 +205,10 @@ Setter for the [`{}::{field_ident}`] field.
                 quote!(
                     impl #const_idents_impl #builder_ident #const_idents_type_input #where_clause {
                         #[doc = #documentation]
-                        pub fn #field_ident (self, #field_ident: #input_type) -> #builder_ident #const_idents_type_output {
-                            let mut #data_field = self.#data_field;
-                            #data_field.#field_ident = #input_value;
+                        pub fn #field_ident (mut self, #field_ident: #input_type) -> #builder_ident #const_idents_type_output {
+                            self.#data_field.#field_ident = #input_value;
                             #builder_ident {
-                                #data_field,
+                                #data_field: self.#data_field,
                             }
                         }
                     }

--- a/const_typed_builder_derive/src/generator/data_generator.rs
+++ b/const_typed_builder_derive/src/generator/data_generator.rs
@@ -155,7 +155,7 @@ impl<'a> DataGenerator<'a> {
             .info
             .field_collection()
             .iter()
-            .filter(|field| field.kind() != &FieldKind::Skipped)
+            .filter(|field| field.kind() != FieldKind::Skipped)
             .map(|field| {
                 let field_ident = field.ident();
                 quote!(#field_ident: None)

--- a/const_typed_builder_derive/src/generator/data_generator.rs
+++ b/const_typed_builder_derive/src/generator/data_generator.rs
@@ -128,7 +128,7 @@ impl<'a> DataGenerator<'a> {
             .iter()
             .map(|field| {
                 let field_ident = field.ident();
-                let tokens = match field.kind() {
+                match field.kind() {
                     FieldKind::Skipped => quote!(#field_ident: None),
                     FieldKind::Mandatory if field.is_option_type() => {
                         quote!(#field_ident: data.#field_ident)
@@ -139,8 +139,7 @@ impl<'a> DataGenerator<'a> {
                     FieldKind::Mandatory => {
                         quote!(#field_ident: data.#field_ident.unwrap())
                     }
-                };
-                tokens
+                }
             })
             .collect()
     }

--- a/const_typed_builder_derive/src/generator/mod.rs
+++ b/const_typed_builder_derive/src/generator/mod.rs
@@ -118,7 +118,7 @@ mod util {
             syn::GenericParam::Const(cnst) => {
                 let ident = &cnst.ident;
                 quote!(#ident)
-            },
+            }
         });
 
         quote!(< #(#generic_idents,)* #(#constants),* >)

--- a/const_typed_builder_derive/src/generator/mod.rs
+++ b/const_typed_builder_derive/src/generator/mod.rs
@@ -107,9 +107,18 @@ mod util {
         generics: &syn::Generics,
     ) -> TokenStream {
         let generic_idents = generics.params.iter().map(|param| match param {
-            syn::GenericParam::Lifetime(lt) => &lt.lifetime.ident,
-            syn::GenericParam::Type(ty) => &ty.ident,
-            syn::GenericParam::Const(cnst) => &cnst.ident,
+            syn::GenericParam::Lifetime(lt) => {
+                let lifetime = &lt.lifetime;
+                quote!(#lifetime)
+            }
+            syn::GenericParam::Type(ty) => {
+                let ident = &ty.ident;
+                quote!(#ident)
+            }
+            syn::GenericParam::Const(cnst) => {
+                let ident = &cnst.ident;
+                quote!(#ident)
+            },
         });
 
         quote!(< #(#generic_idents,)* #(#constants),* >)

--- a/const_typed_builder_derive/src/info/field.rs
+++ b/const_typed_builder_derive/src/info/field.rs
@@ -7,7 +7,7 @@ pub type FieldCollection<'a> = Vec<Field<'a>>;
 
 const CONST_IDENT_PREFIX: &str = "__BUILDER_CONST_";
 
-#[derive(Debug, PartialEq, Eq)]
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
 pub enum FieldKind {
     Optional,
     Skipped,
@@ -15,7 +15,7 @@ pub enum FieldKind {
     Grouped,
 }
 
-#[derive(Debug, PartialEq, Eq)]
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
 pub enum SetterKind {
     Standard,
     Into,
@@ -90,8 +90,8 @@ impl<'a> Field<'a> {
     }
 
     /// Retrieves the kind of the field, which can be Optional, Mandatory, Skipped or Grouped.
-    pub fn kind(&self) -> &FieldKind {
-        &self.kind
+    pub fn kind(&self) -> FieldKind {
+        self.kind
     }
 
     /// Retrieves the index of the field within the struct.
@@ -102,6 +102,10 @@ impl<'a> Field<'a> {
     /// Generates a constant identifier based on the field's index.
     pub fn const_ident(&self) -> syn::Ident {
         format_ident!("{}{}", CONST_IDENT_PREFIX, self.index)
+    }
+
+    pub fn setter_kind(&self) -> SetterKind{
+        self.setter_kind
     }
 
     /// Retrieves the input type for the builder's setter method.

--- a/const_typed_builder_derive/src/info/field.rs
+++ b/const_typed_builder_derive/src/info/field.rs
@@ -31,6 +31,7 @@ pub struct Field<'a> {
     index: usize,
     propagate: bool,
     kind: FieldKind,
+    setter_kind: SetterKind
 }
 
 impl<'a> Field<'a> {
@@ -51,6 +52,7 @@ impl<'a> Field<'a> {
         index: usize,
         kind: FieldKind,
         propagate: bool,
+        setter_kind: SetterKind,
     ) -> Self {
         Self {
             ident,
@@ -58,6 +60,7 @@ impl<'a> Field<'a> {
             ty,
             propagate,
             kind,
+            setter_kind,
         }
     }
 

--- a/const_typed_builder_derive/src/info/field.rs
+++ b/const_typed_builder_derive/src/info/field.rs
@@ -19,8 +19,8 @@ pub enum FieldKind {
 pub enum SetterKind {
     Standard,
     Into,
-    RefMut,
-    AsRef
+    AsMut,
+    AsRef,
 }
 
 /// Represents the information about a struct field used for code generation.
@@ -31,7 +31,7 @@ pub struct Field<'a> {
     index: usize,
     propagate: bool,
     kind: FieldKind,
-    setter_kind: SetterKind
+    setter_kind: SetterKind,
 }
 
 impl<'a> Field<'a> {

--- a/const_typed_builder_derive/src/info/field.rs
+++ b/const_typed_builder_derive/src/info/field.rs
@@ -15,6 +15,14 @@ pub enum FieldKind {
     Grouped,
 }
 
+#[derive(Debug, PartialEq, Eq)]
+pub enum SetterKind {
+    Standard,
+    Into,
+    RefMut,
+    AsRef
+}
+
 /// Represents the information about a struct field used for code generation.
 #[derive(Debug, PartialEq, Eq)]
 pub struct Field<'a> {

--- a/const_typed_builder_derive/src/info/field.rs
+++ b/const_typed_builder_derive/src/info/field.rs
@@ -104,7 +104,7 @@ impl<'a> Field<'a> {
 
 impl<'a> PartialOrd for Field<'a> {
     fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
-        self.index.partial_cmp(&other.index)
+        Some(self.cmp(other))
     }
 }
 

--- a/const_typed_builder_derive/src/info/field.rs
+++ b/const_typed_builder_derive/src/info/field.rs
@@ -18,6 +18,7 @@ pub enum FieldKind {
 #[derive(Debug, PartialEq, Eq, Clone, Copy)]
 pub enum SetterKind {
     Standard,
+    Propagate,
     Into,
     AsMut,
     AsRef,
@@ -29,7 +30,6 @@ pub struct Field<'a> {
     ty: &'a syn::Type,
     ident: &'a syn::Ident,
     index: usize,
-    propagate: bool,
     kind: FieldKind,
     setter_kind: SetterKind,
 }
@@ -51,14 +51,12 @@ impl<'a> Field<'a> {
         ty: &'a syn::Type,
         index: usize,
         kind: FieldKind,
-        propagate: bool,
         setter_kind: SetterKind,
     ) -> Self {
         Self {
             ident,
             index,
             ty,
-            propagate,
             kind,
             setter_kind,
         }
@@ -67,11 +65,6 @@ impl<'a> Field<'a> {
     /// Retrieves the identifier of the field.
     pub fn ident(&self) -> &syn::Ident {
         self.ident
-    }
-
-    /// Checks if the field's attributes indicate propagation.
-    pub fn propagate(&self) -> bool {
-        self.propagate
     }
 
     /// Checks if the field's type is an Option.
@@ -104,25 +97,8 @@ impl<'a> Field<'a> {
         format_ident!("{}{}", CONST_IDENT_PREFIX, self.index)
     }
 
-    pub fn setter_kind(&self) -> SetterKind{
+    pub fn setter_kind(&self) -> SetterKind {
         self.setter_kind
-    }
-
-    /// Retrieves the input type for the builder's setter method.
-    pub fn setter_input_type(&self) -> Option<&syn::Type> {
-        match self.kind() {
-            FieldKind::Optional => Some(self.ty()),
-            FieldKind::Mandatory if self.is_option_type() => Some(self.inner_type().expect(
-                "Couldn't read inner type of option, even though it's seen as an Option type",
-            )),
-            FieldKind::Mandatory => Some(self.ty()),
-            FieldKind::Grouped => {
-                Some(self.inner_type().expect(
-                    "Couldn't read inner type of option, even though it's marked as grouped",
-                ))
-            }
-            FieldKind::Skipped => None,
-        }
     }
 }
 

--- a/const_typed_builder_derive/src/info/mod.rs
+++ b/const_typed_builder_derive/src/info/mod.rs
@@ -3,5 +3,5 @@ mod field;
 mod group;
 
 pub use container::{Container, SolverKind};
-pub use field::{Field, FieldCollection, FieldKind, TrackedField, TrackedFieldKind};
+pub use field::{Field, FieldCollection, FieldKind, TrackedField, TrackedFieldKind, SetterKind};
 pub use group::{Group, GroupCollection, GroupType};

--- a/const_typed_builder_derive/src/info/mod.rs
+++ b/const_typed_builder_derive/src/info/mod.rs
@@ -3,5 +3,5 @@ mod field;
 mod group;
 
 pub use container::{Container, SolverKind};
-pub use field::{Field, FieldCollection, FieldKind, TrackedField, TrackedFieldKind, SetterKind};
+pub use field::{Field, FieldCollection, FieldKind, SetterKind, TrackedField, TrackedFieldKind};
 pub use group::{Group, GroupCollection, GroupType};

--- a/const_typed_builder_derive/src/parser/container_parser.rs
+++ b/const_typed_builder_derive/src/parser/container_parser.rs
@@ -9,6 +9,7 @@ use proc_macro_error::{emit_call_site_error, emit_error, emit_warning};
 #[derive(Debug)]
 pub struct ContainerParser {
     assume_mandatory: bool,
+    assume_into: bool,
     /// A map of group names to their respective `GroupInfo`.
     groups: GroupCollection,
     /// The solver used to find all possible valid combinations for the groups
@@ -120,6 +121,9 @@ impl ContainerParser {
                 symbol::ASSUME_MANDATORY => {
                     self.assume_mandatory = true;
                 }
+                symbol::INTO => {
+                    self.assume_into = true;
+                }
                 symbol::SOLVER => {
                     let syn::ExprPath { path, .. } = meta.value()?.parse()?;
                     match (&path.require_ident()?.to_string()).into() {
@@ -180,7 +184,7 @@ impl ContainerParser {
                     .ident
                     .as_ref()
                     .expect("FieldsNamed should have an ident");
-                FieldParser::new(index, self.assume_mandatory, &mut self.groups).parse(ident, field)
+                FieldParser::new(index, self.assume_mandatory, self.assume_into, &mut self.groups).parse(ident, field)
             })
             .collect::<Vec<_>>()
     }
@@ -190,6 +194,7 @@ impl Default for ContainerParser {
     fn default() -> Self {
         ContainerParser {
             assume_mandatory: false,
+            assume_into: false,
             groups: GroupCollection::new(),
             solver_kind: SolverKind::BruteForce,
         }

--- a/const_typed_builder_derive/src/parser/container_parser.rs
+++ b/const_typed_builder_derive/src/parser/container_parser.rs
@@ -184,7 +184,13 @@ impl ContainerParser {
                     .ident
                     .as_ref()
                     .expect("FieldsNamed should have an ident");
-                FieldParser::new(index, self.assume_mandatory, self.assume_into, &mut self.groups).parse(ident, field)
+                FieldParser::new(
+                    index,
+                    self.assume_mandatory,
+                    self.assume_into,
+                    &mut self.groups,
+                )
+                .parse(ident, field)
             })
             .collect::<Vec<_>>()
     }

--- a/const_typed_builder_derive/src/parser/field_parser.rs
+++ b/const_typed_builder_derive/src/parser/field_parser.rs
@@ -10,7 +10,6 @@ use proc_macro_error::{emit_error, emit_warning};
 pub struct FieldParser<'parser> {
     kind: Option<FieldKind>,
     setter_kind: Option<SetterKind>,
-    propagate: bool,
     index: usize,
     assume_mandatory: bool,
     group_collection: &'parser mut GroupCollection,
@@ -25,7 +24,6 @@ impl<'parser> FieldParser<'parser> {
         Self {
             kind: None,
             setter_kind: None,
-            propagate: false,
             index,
             assume_mandatory,
             group_collection,
@@ -60,7 +58,6 @@ impl<'parser> FieldParser<'parser> {
             ty,
             self.index,
             self.kind.unwrap(),
-            self.propagate,
             self.setter_kind.unwrap(),
         )
     }
@@ -138,8 +135,11 @@ impl<'parser> FieldParser<'parser> {
                 symbol::MANDATORY => self.handle_attribute_mandatory(path_ident),
                 symbol::OPTIONAL => self.handle_attribute_optional(path_ident),
                 symbol::GROUP => self.handle_attribute_group(&meta),
-                symbol::PROPAGATE => self.propagate = true,
-                symbol::SETTER => self.handle_attribute_setter(&meta)?,
+                symbol::PROPAGATE => self.handle_setter_kind(SetterKind::Propagate, path_ident),
+                symbol::ASREF => self.handle_setter_kind(SetterKind::AsRef, path_ident),
+                symbol::ASMUT => self.handle_setter_kind(SetterKind::AsMut, path_ident),
+                symbol::INTO => self.handle_setter_kind(SetterKind::Into, path_ident),
+                symbol::STANDARD => self.handle_setter_kind(SetterKind::Standard, path_ident),
                 _ => emit_error!(&attr.meta, "Unknown attribute"),
             }
             Ok(())
@@ -152,22 +152,12 @@ impl<'parser> FieldParser<'parser> {
         })
     }
 
-    fn handle_attribute_setter(&mut self, meta: &syn::meta::ParseNestedMeta) -> syn::Result<()> {
-        let setter_ident: syn::Ident = meta.value()?.parse()?;
+    fn handle_setter_kind(&mut self, kind: SetterKind, ident: &syn::Ident) {
         if self.setter_kind.is_some() {
-            emit_error!(setter_ident, "Setter type defined multiple times");
-            return Ok(());
+            emit_error!(ident, "Setter type defined multiple times");
+        } else {
+            self.setter_kind = Some(kind);
         }
-        match (&setter_ident.to_string()).into() {
-            symbol::ASREF => self.setter_kind = Some(SetterKind::AsRef),
-            symbol::ASMUT => self.setter_kind = Some(SetterKind::AsMut),
-            symbol::INTO => self.setter_kind = Some(SetterKind::Into),
-            symbol::STANDARD => self.setter_kind = Some(SetterKind::Standard),
-            _ => {
-                emit_error!(setter_ident, "Unknown setter type");
-            }
-        }
-        Ok(())
     }
 
     fn handle_attribute_skip(&mut self, ident: &syn::Ident) {

--- a/const_typed_builder_derive/src/parser/field_parser.rs
+++ b/const_typed_builder_derive/src/parser/field_parser.rs
@@ -12,6 +12,7 @@ pub struct FieldParser<'parser> {
     setter_kind: Option<SetterKind>,
     index: usize,
     assume_mandatory: bool,
+    assume_into: bool,
     group_collection: &'parser mut GroupCollection,
 }
 
@@ -19,6 +20,7 @@ impl<'parser> FieldParser<'parser> {
     pub fn new(
         index: usize,
         assume_mandatory: bool,
+        assume_into: bool,
         group_collection: &'parser mut GroupCollection,
     ) -> Self {
         Self {
@@ -26,6 +28,7 @@ impl<'parser> FieldParser<'parser> {
             setter_kind: None,
             index,
             assume_mandatory,
+            assume_into,
             group_collection,
         }
     }
@@ -50,7 +53,11 @@ impl<'parser> FieldParser<'parser> {
         }
 
         if self.setter_kind.is_none() {
-            self.setter_kind = Some(SetterKind::Standard);
+            self.setter_kind = if self.assume_into {
+                Some(SetterKind::Into)
+            } else {
+                Some(SetterKind::Standard)
+            };
         }
 
         Field::new(

--- a/const_typed_builder_derive/src/parser/field_parser.rs
+++ b/const_typed_builder_derive/src/parser/field_parser.rs
@@ -9,6 +9,7 @@ use proc_macro_error::{emit_error, emit_warning};
 #[derive(Debug)]
 pub struct FieldParser<'parser> {
     kind: Option<FieldKind>,
+    setter_kind: Option<SetterKind>,
     propagate: bool,
     index: usize,
     assume_mandatory: bool,
@@ -23,6 +24,7 @@ impl<'parser> FieldParser<'parser> {
     ) -> Self {
         Self {
             kind: None,
+            setter_kind: None,
             propagate: false,
             index,
             assume_mandatory,
@@ -46,10 +48,21 @@ impl<'parser> FieldParser<'parser> {
                 Some(FieldKind::Mandatory)
             } else {
                 Some(FieldKind::Optional)
-            }
+            };
         }
 
-        Field::new(ident, ty, self.index, self.kind.unwrap(), self.propagate, SetterKind::Standard)
+        if self.setter_kind.is_none() {
+            self.setter_kind = Some(SetterKind::Standard);
+        }
+
+        Field::new(
+            ident,
+            ty,
+            self.index,
+            self.kind.unwrap(),
+            self.propagate,
+            self.setter_kind.unwrap(),
+        )
     }
 
     /// Handles the parsing and processing of a builder attribute applied to a field.
@@ -124,18 +137,37 @@ impl<'parser> FieldParser<'parser> {
                 symbol::SKIP => self.handle_attribute_skip(path_ident),
                 symbol::MANDATORY => self.handle_attribute_mandatory(path_ident),
                 symbol::OPTIONAL => self.handle_attribute_optional(path_ident),
-                symbol::GROUP => self.handle_attribute_group(path_ident, &meta),
+                symbol::GROUP => self.handle_attribute_group(&meta),
                 symbol::PROPAGATE => self.propagate = true,
+                symbol::SETTER => self.handle_attribute_setter(&meta)?,
                 _ => emit_error!(&attr.meta, "Unknown attribute"),
             }
             Ok(())
         })
         .unwrap_or_else(|err| {
             emit_error!(
-                &attr.meta, "Unknown error";
+                &attr.meta, "An error occured while parsing this attribute";
                 note = err
             )
         })
+    }
+
+    fn handle_attribute_setter(&mut self, meta: &syn::meta::ParseNestedMeta) -> syn::Result<()> {
+        let setter_ident: syn::Ident = meta.value()?.parse()?;
+        if self.setter_kind.is_some() {
+            emit_error!(setter_ident, "Setter type defined multiple times");
+            return Ok(());
+        }
+        match (&setter_ident.to_string()).into() {
+            symbol::ASREF => self.setter_kind = Some(SetterKind::AsRef),
+            symbol::ASMUT => self.setter_kind = Some(SetterKind::AsMut),
+            symbol::INTO => self.setter_kind = Some(SetterKind::Into),
+            symbol::STANDARD => self.setter_kind = Some(SetterKind::Standard),
+            _ => {
+                emit_error!(setter_ident, "Unknown setter type");
+            }
+        }
+        Ok(())
     }
 
     fn handle_attribute_skip(&mut self, ident: &syn::Ident) {
@@ -201,19 +233,19 @@ impl<'parser> FieldParser<'parser> {
         }
     }
 
-    fn handle_attribute_group(&mut self, ident: &syn::Ident, meta: &syn::meta::ParseNestedMeta) {
+    fn handle_attribute_group(&mut self, meta: &syn::meta::ParseNestedMeta) {
         match self.kind {
             None => self.kind = Some(FieldKind::Grouped),
             Some(FieldKind::Optional) => emit_error!(
-                ident, "Can't define field as part of a group as its already defined as optional";
+                meta.path, "Can't define field as part of a group as its already defined as optional";
                 hint = "Remove either types of attribute from this field"
             ),
             Some(FieldKind::Skipped) => emit_error!(
-                ident, "Can't define field as as part of a group as its already defined as skipped";
+                meta.path, "Can't define field as as part of a group as its already defined as skipped";
                 hint = "Remove either types of attribute from this field"
             ),
             Some(FieldKind::Mandatory) => emit_error!(
-                ident, "Can't define field as as part of a group as its already defined as mandatory";
+                meta.path, "Can't define field as as part of a group as its already defined as mandatory";
                 hint = "Remove either types of attribute from this field"
             ),
             Some(FieldKind::Grouped) => {}
@@ -233,7 +265,7 @@ impl<'parser> FieldParser<'parser> {
             }
             Err(err) => {
                 emit_error!(
-                    ident, "Group name not specified correctly";
+                    meta.path, "Group name not specified correctly";
                     help = "Try defining it like #[{}(foo)]", symbol::BUILDER;
                     note = err
                 );

--- a/const_typed_builder_derive/src/parser/field_parser.rs
+++ b/const_typed_builder_derive/src/parser/field_parser.rs
@@ -1,5 +1,5 @@
 use crate::{
-    info::{Field, FieldKind, GroupCollection},
+    info::{Field, FieldKind, GroupCollection, SetterKind},
     symbol,
     util::is_option,
 };
@@ -49,7 +49,7 @@ impl<'parser> FieldParser<'parser> {
             }
         }
 
-        Field::new(ident, ty, self.index, self.kind.unwrap(), self.propagate)
+        Field::new(ident, ty, self.index, self.kind.unwrap(), self.propagate, SetterKind::Standard)
     }
 
     /// Handles the parsing and processing of a builder attribute applied to a field.

--- a/const_typed_builder_derive/src/symbol.rs
+++ b/const_typed_builder_derive/src/symbol.rs
@@ -34,6 +34,16 @@ pub const BRUTE_FORCE: Symbol = Symbol("brute_force");
 pub const COMPILER: Symbol = Symbol("compiler");
 /// Constant representing the `skip` symbol.
 pub const SKIP: Symbol = Symbol("skip");
+/// Constant representing the `setter` symbol.
+pub const SETTER: Symbol = Symbol("setter");
+/// Constant representing the `setter` symbol.
+pub const INTO: Symbol = Symbol("into");
+/// Constant representing the `setter` symbol.
+pub const ASREF: Symbol = Symbol("asref");
+/// Constant representing the `setter` symbol.
+pub const ASMUT: Symbol = Symbol("asmut");
+/// Constant representing the `setter` symbol.
+pub const STANDARD: Symbol = Symbol("standard");
 
 impl<'a> From<&'a String> for Symbol<'a> {
     fn from(value: &'a String) -> Self {

--- a/const_typed_builder_derive/src/symbol.rs
+++ b/const_typed_builder_derive/src/symbol.rs
@@ -35,8 +35,6 @@ pub const COMPILER: Symbol = Symbol("compiler");
 /// Constant representing the `skip` symbol.
 pub const SKIP: Symbol = Symbol("skip");
 /// Constant representing the `setter` symbol.
-pub const SETTER: Symbol = Symbol("setter");
-/// Constant representing the `setter` symbol.
 pub const INTO: Symbol = Symbol("into");
 /// Constant representing the `setter` symbol.
 pub const ASREF: Symbol = Symbol("asref");

--- a/const_typed_builder_test/src/lib.rs
+++ b/const_typed_builder_test/src/lib.rs
@@ -730,6 +730,41 @@ mod test {
     }
 
     #[test]
+    fn assume_into() {
+        #[derive(Debug, PartialEq, Builder)]
+        #[builder(into)]
+        pub struct Foo {
+            bar: String,
+        }
+        let expected = Foo {
+            bar: "Hello world!".to_string(),
+        };
+        let foo = Foo::builder().bar("Hello world!").build();
+        assert_eq!(foo, expected);
+    }
+
+    #[test]
+    fn into_other_strct() {
+        #[derive(Debug, PartialEq)]
+        pub struct MyStruct {
+            my_str: String
+        }
+        impl MyStruct {
+            fn new(my_str: String) -> Self { Self { my_str } }
+        }
+        #[derive(Debug, PartialEq, Builder)]
+        #[builder(into)]
+        pub struct Foo {
+            bar: MyStruct,
+        }
+        let expected = Foo {
+            bar: MyStruct { my_str: "Hello world!".to_string() },
+        };
+        let foo = Foo::builder().bar(MyStruct::new("Hello world!".to_string())).build();
+        assert_eq!(foo, expected);
+    }
+
+    #[test]
     fn asref() {
         #[derive(Debug, PartialEq, Builder)]
         pub struct Foo<'a> {
@@ -791,10 +826,10 @@ mod test {
         let mut m_str_clone = m_str.clone();
 
         let expected = Foo {
-            bar: Some(&mut m_str_clone),
+            bar: Some(&mut m_str),
         };
         
-        let foo = Foo::builder().bar(Some(&mut m_str)).build();
+        let foo = Foo::builder().bar(Some(&mut m_str_clone)).build();
         assert_eq!(foo, expected);
     }
 }

--- a/const_typed_builder_test/src/lib.rs
+++ b/const_typed_builder_test/src/lib.rs
@@ -689,6 +689,21 @@ mod test {
     }
 
     #[test]
+    fn mut_reference() {
+        #[derive(Debug, PartialEq, Builder)]
+        pub struct Foo<'a> {
+            bar: &'a mut str,
+        }
+        let mut my_str = "Hello world!".to_string();
+        let mut my_str_clone = my_str.clone();
+        let expected = Foo {
+            bar: my_str.as_mut_str(),
+        };
+        let foo = Foo::builder().bar(my_str_clone.as_mut_str()).build();
+        assert_eq!(foo, expected);
+    }
+
+    #[test]
     fn const_generic() {
         #[derive(Debug, PartialEq, Builder)]
         pub struct Foo<const SIZE: usize> {

--- a/const_typed_builder_test/src/lib.rs
+++ b/const_typed_builder_test/src/lib.rs
@@ -690,13 +690,11 @@ mod test {
 
     #[test]
     fn const_generic() {
-    #[derive(Debug, PartialEq, Builder)]
+        #[derive(Debug, PartialEq, Builder)]
         pub struct Foo<const SIZE: usize> {
             bar: [usize; SIZE],
         }
-        let expected = Foo {
-            bar: [1, 2, 3],
-        };
+        let expected = Foo { bar: [1, 2, 3] };
         let foo = Foo::builder().bar([1, 2, 3]).build();
         assert_eq!(foo, expected);
     }
@@ -747,10 +745,12 @@ mod test {
     fn into_other_strct() {
         #[derive(Debug, PartialEq)]
         pub struct MyStruct {
-            my_str: String
+            my_str: String,
         }
         impl MyStruct {
-            fn new(my_str: String) -> Self { Self { my_str } }
+            fn new(my_str: String) -> Self {
+                Self { my_str }
+            }
         }
         #[derive(Debug, PartialEq, Builder)]
         #[builder(into)]
@@ -758,9 +758,13 @@ mod test {
             bar: MyStruct,
         }
         let expected = Foo {
-            bar: MyStruct { my_str: "Hello world!".to_string() },
+            bar: MyStruct {
+                my_str: "Hello world!".to_string(),
+            },
         };
-        let foo = Foo::builder().bar(MyStruct::new("Hello world!".to_string())).build();
+        let foo = Foo::builder()
+            .bar(MyStruct::new("Hello world!".to_string()))
+            .build();
         assert_eq!(foo, expected);
     }
 
@@ -773,10 +777,8 @@ mod test {
         }
         let m_str = "Hello world!".to_string();
 
-        let expected = Foo {
-            bar: &m_str,
-        };
-        
+        let expected = Foo { bar: &m_str };
+
         let foo = Foo::builder().bar(&m_str).build();
         assert_eq!(foo, expected);
     }
@@ -793,7 +795,7 @@ mod test {
         let expected = Foo {
             bar: Some(m_str.as_str()),
         };
-        
+
         let foo = Foo::builder().bar(Some(&m_str)).build();
         assert_eq!(foo, expected);
     }
@@ -810,7 +812,7 @@ mod test {
         let expected = Foo {
             bar: &mut m_str.clone(),
         };
-        
+
         let foo = Foo::builder().bar(&mut m_str).build();
         assert_eq!(foo, expected);
     }
@@ -828,7 +830,7 @@ mod test {
         let expected = Foo {
             bar: Some(&mut m_str),
         };
-        
+
         let foo = Foo::builder().bar(Some(&mut m_str_clone)).build();
         assert_eq!(foo, expected);
     }

--- a/const_typed_builder_test/src/lib.rs
+++ b/const_typed_builder_test/src/lib.rs
@@ -674,4 +674,127 @@ mod test {
         let foo = Foo::builder().bar("Hello world!".to_string()).build();
         assert_eq!(foo, expected);
     }
+
+    #[test]
+    fn reference() {
+        #[derive(Debug, PartialEq, Builder)]
+        pub struct Foo<'a> {
+            bar: &'a str,
+        }
+        let expected = Foo {
+            bar: "Hello world!",
+        };
+        let foo = Foo::builder().bar("Hello world!").build();
+        assert_eq!(foo, expected);
+    }
+
+    #[test]
+    fn const_generic() {
+    #[derive(Debug, PartialEq, Builder)]
+        pub struct Foo<const SIZE: usize> {
+            bar: [usize; SIZE],
+        }
+        let expected = Foo {
+            bar: [1, 2, 3],
+        };
+        let foo = Foo::builder().bar([1, 2, 3]).build();
+        assert_eq!(foo, expected);
+    }
+
+    #[test]
+    fn into() {
+        #[derive(Debug, PartialEq, Builder)]
+        pub struct Foo {
+            #[builder(into)]
+            bar: String,
+        }
+        let expected = Foo {
+            bar: "Hello world!".to_string(),
+        };
+        let foo = Foo::builder().bar("Hello world!").build();
+        assert_eq!(foo, expected);
+    }
+
+    #[test]
+    fn into_optional() {
+        #[derive(Debug, PartialEq, Builder)]
+        pub struct Foo {
+            #[builder(into)]
+            bar: Option<String>,
+        }
+        let expected = Foo {
+            bar: Some("Hello world!".to_string()),
+        };
+        let foo = Foo::builder().bar(Some("Hello world!")).build();
+        assert_eq!(foo, expected);
+    }
+
+    #[test]
+    fn asref() {
+        #[derive(Debug, PartialEq, Builder)]
+        pub struct Foo<'a> {
+            #[builder(asref)]
+            bar: &'a str,
+        }
+        let m_str = "Hello world!".to_string();
+
+        let expected = Foo {
+            bar: &m_str,
+        };
+        
+        let foo = Foo::builder().bar(&m_str).build();
+        assert_eq!(foo, expected);
+    }
+
+    #[test]
+    fn asref_optional() {
+        #[derive(Debug, PartialEq, Builder)]
+        pub struct Foo<'a> {
+            #[builder(asref)]
+            bar: Option<&'a str>,
+        }
+        let m_str = "Hello world!".to_string();
+
+        let expected = Foo {
+            bar: Some(m_str.as_str()),
+        };
+        
+        let foo = Foo::builder().bar(Some(&m_str)).build();
+        assert_eq!(foo, expected);
+    }
+
+    #[test]
+    fn asmut() {
+        #[derive(Debug, PartialEq, Builder)]
+        pub struct Foo<'a> {
+            #[builder(asmut)]
+            bar: &'a mut str,
+        }
+        let mut m_str = "Hello world!".to_string();
+
+        let expected = Foo {
+            bar: &mut m_str.clone(),
+        };
+        
+        let foo = Foo::builder().bar(&mut m_str).build();
+        assert_eq!(foo, expected);
+    }
+
+    #[test]
+    fn asmut_optional() {
+        #[derive(Debug, PartialEq, Builder)]
+        pub struct Foo<'a> {
+            #[builder(asmut)]
+            bar: Option<&'a mut str>,
+        }
+        let mut m_str = "Hello world!".to_string();
+        let mut m_str_clone = m_str.clone();
+
+        let expected = Foo {
+            bar: Some(&mut m_str_clone),
+        };
+        
+        let foo = Foo::builder().bar(Some(&mut m_str)).build();
+        assert_eq!(foo, expected);
+    }
 }

